rholang/examples/linking/packages/MakeBrandPair.rho

```diff
@@ -0,0 +1,18 @@
+//based on sealer/unsealer pattern in http://erights.org/elib/capability/ode/ode-capabilities.html
+export MakeBrandPair in {
+  contract MakeBrandPair(return) = {
+    new seal, unseal, key in {
+      contract seal(data, return) = {
+        new sealedData in {
+          contract sealedData(input, return) = {
+            match input == key with true => { return(data) }
+            _ => { return(Nil) }
+          } |
+          return(sealedData)
+        }
+      } |
+      contract unseal(sealedData, return) = { sealedData(key, return) } |
+      return([seal, unseal])
```

Comment:
This only works because we haven't implemented the ability to destructure processes yet:
```
new ret in {
  MakeBrandPair(ret) |
  for ([seal, unseal] <- ret) {
    delegate!(unseal)
  }
} |
for (seal <- delegate) {
  match seal with
    // All the names here are bound except key,
    // which acts as a binder and grabs the actual
    // unforgeable name used in seal.
    {
      new sealedData in {
        contract sealedData(input, return) = {
          match input == key with true => { return(data) }
          _ => { return(Nil) }
        } |
        return(sealedData)
     }
   } => leak!(key)
}
```
When we get that ability, the seal/unseal processes will need to include the `opaque` process to prevent clients from destructuring them.

Reply:
Maybe add a comment to that effect so we don't forget?

Reply:
That's not true with what @birchmd has written. For each sealer/unsealer pair, there's a single seal/unseal contract, and the name of that contract gets passed around. Opaque would allow us to pass the code around, rather than having a global lock on unsealing.

Reply:
I'm not sure what you mean. There is a single constructor of sealer/unsealer pairs, but that constructor gives a unique, unforgable key to each pair it creates. The problem that @pyrocto pointed out was that you could extract the key by desconstructing the output of the sealer (which is a new contract that will return data only when it gets the right key). So making that output opaque prevents that problem. Destructuring the unsealer to get the key doesn't matter because you already have the unsealer, so you have permission to extract the data anyways.